File: LeetCode-Cpp/Design Add and SearchWordsDataStructure.cpp
class WordDictionary {
private:
    struct TrieNode {
        bool isEnd;  // bir kelimenin sonunu işaretler
        unordered_map<char, TrieNode*> children;
        TrieNode() : isEnd(false) {}
    };
    
    TrieNode* root;

    bool dfs(string &word, int index, TrieNode* node) {
        if (index == word.size()) {
            return node->isEnd;  // kelimenin sonuna ulaştık mı?
        }
        
        char c = word[index];
        if (c == '.') {
            // '.' → tüm çocukları denemeliyiz
            for (auto &p : node->children) {
                if (dfs(word, index + 1, p.second)) {
                    return true;
                }
            }
            return false;
        } else {
            if (!node->children.count(c)) return false;
            return dfs(word, index + 1, node->children[c]);
        }
    }
public:
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* node = root;
        for (char c : word) {
            if (!node->children.count(c)) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        return dfs(word, 0, root);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: LeetCode-Python/DesignAddandSearchWordsDataStructure.py
class TrieNode:
    def __init__(self):
        self.children = {}   # harf → TrieNode
        self.isEnd = False   # kelimenin bitişini işaretler

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True

    def search(self, word: str) -> bool:
        def dfs(index: int, node: TrieNode) -> bool:
            if index == len(word):
                return node.isEnd
            c = word[index]
            if c == ".":
                # '.' → tüm çocukları dene
                for child in node.children.values():
                    if dfs(index + 1, child):
                        return True
                return False
            else:
                if c not in node.children:
                    return False
                return dfs(index + 1, node.children[c])
        
        return dfs(0, self.root)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
